src/nodeattr/nodeattr.c

```diff
@@ -1000,6 +971,70 @@ _hash_attrval(hash_t hattr, char *node, char *attr, char *val)
     }
 }
 
+static void 
+list_attrs(genders_t gp, char *node)
+{
+    char **attrs, **vals;
+    char *anode = NULL;
+    int numattrs, len, vlen, count, i;
+    hash_t hattrval = NULL;
+    hostlist_t nodelist;
+
+    if ((len = genders_attrlist_create(gp, &attrs)) < 0)
+        _gend_error_exit(gp, "genders_attrlist_create");
+    if ((vlen = genders_vallist_create(gp, &vals)) < 0)
+        _gend_error_exit(gp, "genders_vallist_create");
+    if ((numattrs = genders_getnumattrs(gp)) < 0)
+        _gend_error_exit(gp, "genders_getnumattrs");
+    if (!(hattrval = hash_create((numattrs + 1)*4,
+                            (hash_key_f)hash_key_string,
+                            (hash_cmp_f)strcmp,
+                            _hosts_data_del))) {
+        fprintf(stderr, "hash_create: %s\n", strerror(errno));
+        exit(1);
+    }
+
+    /* a single node or a list of nodes */
+    if (node) {
+        nodelist = hostlist_create(node);
+        while ((anode = hostlist_pop(nodelist))) {
+            if ((count = genders_getattr(gp, attrs, vals, len, anode)) < 0)
+                _gend_error_exit(gp, "genders_getattr");
+            for (i = 0; i < count; i++)
+                _hash_attrval(hattrval, anode, attrs[i], vals[i]);
+            free(anode);
+        }
+        
+        if (hash_for_each(hattrval, _print_key, NULL) < 0) {
+            fprintf(stderr, "hash_for_each: %s\n", strerror(errno));
+            exit(1);
+        }
+        hostlist_destroy(nodelist);
+    /* all nodes */
+    } else {
+        if ((count = genders_getattr_all(gp, attrs, len)) < 0)
+            _gend_error_exit(gp, "genders_getattr_all");
+        for (i = 0; i < count; i++)
+            if (node && strlen(vals[i]) > 0)
+                printf("%s=%s\n", attrs[i], vals[i]);
+            else
+                printf("%s\n", attrs[i]);
+    }

```

Comment:
logic simplification, if you're in the "all nodes" case, there should be no vals to ever print out?

Reply:
yes, thanks!